crates/gateway-types/src/reply.rs

```diff
@@ -1066,14 +1063,77 @@ pub mod transaction {
         pub contract_address: ContractAddress,
         pub entry_point_selector: EntryPoint,
         // FIXME: remove once starkware fixes their gateway bug which was missing this field.
-        #[serde(default = "l1_handler_default_nonce")]
+        #[serde(default)]
         pub nonce: TransactionNonce,
         pub calldata: Vec<CallParam>,
         pub transaction_hash: TransactionHash,
         #[serde_as(as = "TransactionVersionAsHexStr")]
         pub version: TransactionVersion,
     }
 
+    impl L1HandlerTransaction {
+        pub fn calculate_message_hash(&self) -> H256 {
+            use sha3::{Digest, Keccak256};
+
+            let Some((from_address, payload)) = self.calldata.split_first() else {
+                // This would indicate a pretty severe error in the L1 transaction.
+                // But since we haven't encoded this during serialization, this could in
+                // theory mess us up here.
+                //
+                // We should incorporate this into the deserialization instead. Returning an
+                // error here is unergonomic and far too late.
+                return H256::zero();
+            };
```

Comment:
This is a bit awkward; but I don't really want to pollute everything with an error in-case the transaction is completely invalid. In reality we should actually handle this assumption when deserializing already.